js/core/options/index.js

```diff
@@ -0,0 +1,223 @@
+import createCallBack from '../utils/callbacks';
+import { isFunction, isObject, type } from '../utils/type';
+import { equalByValue } from '../utils/common';
+import { OptionManager } from './option_manager';
+import { clone } from '../utils/object';
+import { getFieldName, getParentName, convertRulesToOptions } from './utils';
+
+export class Options {
+    constructor(options, defaultOptions, optionsByReference, deprecatedOptions) {
+        this._deprecatedCallbacks = createCallBack({ syncStrategy: true });
+
+        this._default = defaultOptions;
+        this._deprecated = deprecatedOptions;
+
+        this._cachedDeprecateNames = [];
+
+        this._optionManager = new OptionManager(
+            options,
+            optionsByReference
+        );
+        this._optionManager.onRelevantNamesPrepared((options, name, value) => this._setRelevantNames(options, name, value));
+
+        this._rules = [];
+    }
+
+    get _options() {
+        return this._optionManager._options;
+    }
+
+    get _deprecateNames() {
+        if(!this._cachedDeprecateNames.length) {
+            for(const optionName in this._deprecated) {
+                this._cachedDeprecateNames.push(optionName);
+            }
+        }
+
+        return this._cachedDeprecateNames;
```

Comment:
Seems like `this._cachedDeprecateNames.push(optionName);` we should do in the constructor